### ISSUE TITLE
this.rerender() undefined

### ISSUE THREAD

**Issue by afincato** [choojs/nanocomponent#85]
Hi!

Noob here so I might just be missing something obvious or complicating things for no reason.

I am in a situation where I have to create a choo component that sets its `width` and `height` values by first loading an image as a child node, and then using that image to defines `width` and `height` of the root node. That works.

The problem comes when I add a `resize` `EventListener` and try to `rerender()` the component. If I log the function `this.rerender()` I always get undefined.

Some relevant code ([file here):](https://github.com/afincato/face-the-interface/blob/flipcard_comp/app/components/card.js)

```
load (el) {
  const resize = () => {
    if (el.firstChild.firstChild.nodeName.toLowerCase() === 'img') {
      this.width = el.firstElementChild.clientWidth
      this.height = (el.firstElementChild.firstChild.dataset.height / el.firstElementChild.firstChild.dataset.width) * el.firstElementChild.firstChild.clientWidth
    } else {
      this.width = el.firstElementChild.firstChild.clientWidth
      this.height = (el.firstElementChild.firstChild.firstChild.dataset.height / el.firstElementChild.firstChild.firstChild.dataset.width) * el.firstElementChild.firstChild.firstChild.clientWidth
    }

    this.rerender.bind(this)()
    console.log( this.rerender.bind(this)() )
}

resize()

window.addEventListener('resize', function () {
  resize()
})

}

update (el) {
  return true
}
```

and how I create the component ([file here](https://github.com/afincato/face-the-interface/blob/flipcard_comp/app/views/blog.js)):

```
function entry (state, emit) {
  return ov(page[0].children).map(function (item, i) {

  var card = state.cache(Card, i)
    return card.render(state, emit, item, i, state.cards[i])
  })
}
```

Could it be that `state.cache` does not see that I am changing the `width` and `height` values when resizing the page, so... I should be explicit about that?

Thanks,
af

**Comment by tornqvist** [comment]
Firstly, you've got the constructor arguments wrong in your `Card` component. They should be `id`, `state` and `emit`, and id should preferably be forwarded to `super`. But that shouldn't cause any problem in your use case, since you overwrite them in `createElement`.

Secondly, `this.rerender()` isn't supposed to return anything (hence `undefined`), it just rerenders the element in place. Also, there's no need for the `bind` in your use case.

Is the issue that you component doesn't rerender on resize, or was it just that `rerender` returns undefined?

**Comment by afincato** [comment]
Thanks for the tips!

My issue is that the component does not rerender on resize, but from what you are saying now, rerender is not supposed to do that?

**Comment by afincato** [comment]
Blessed the conversation I had with you guys!

In an attempt to try another solution, I replaced

```
this.width = el.firstElementChild.clientWidth
this.height = (el.firstElementChild.firstChild.dataset.height / el.firstElementChild.firstChild.dataset.width) * el.firstElementChild.firstChild.clientWidth
```

with

```
this.width = el.firstElementChild.clientWidth
this.height = (el.firstElementChild.firstChild.dataset.height / el.firstElementChild.firstChild.dataset.width) * this.width
```

so using the cached `this.width` value from the line above, and the resize began working correctly!

In order to better understand, what's the logic behind this? That I'm using a cached value?

**Comment by bcomnes** [comment]
Hard to parse exactly whats going on without more context, but using a cached value is sometimes necessary depending on the order of things.

**Comment by afincato** [comment]
My hunch without knowing how choo and nanocomponent work, is that maybe since I am using `state.cache()` to create the component, choo was not finding any reason to trigger a diff and therefore re-render the component?

```
function entry (state, emit) {
      return ov(page[0].children).map(function (item, i) {

        var card = state.cache(Card, i)
        return card.render(state, emit, item, i, state.cards[i])
      })
}
``` 
([ref](https://github.com/afincato/face-the-interface/blob/master/app/views/blog.js#L43))

Anyway, glad this worked out after 5 weeks, and thanks for the chat, very helpful!

**Comment by goto-bus-stop** [comment]
In your first line you're using:

```js
el.firstElementChild.clientWidth
```

as the `this.width` value. But in the second line you're multiplying by
```js
el.firstElementChild.firstChild.clientWidth
```

that's probably what's causing the difference

**Comment by afincato** [comment]
Ahah, yeah! You're actually right — just tested 😅
